pkg/helm/client.go

```diff
@@ -0,0 +1,247 @@
+package helm
+
+import (
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/blang/semver"
+	"github.com/ghodss/yaml"
+	"github.com/pkg/errors"
+)
+
+var (
+	validHelmSchemes = map[string]bool{
+		"http":  true,
+		"https": true,
+	}
+)
+
+// RepositoryClient is a client for retrieving Helm repository contents. The structure
+// is loosely documented at https://github.com/kubernetes/helm/blob/master/docs/chart_repository.md.
+type RepositoryClient interface {
+	// Repository returns the contents of a Helm repository.
+	Repository() (*Repository, error)
+	// Chart returns a Chart with a given name and version. If the version is blank, it returns
+	// the latest version.
+	Chart(name, version string) (*RepositoryChart, error)
+	// Fetch fetches URIs for a chart.
+	Fetch(uri string) (io.ReadCloser, error)
+}
+
+// Repository is metadata describing the contents of a Helm repository.
+type Repository struct {
+	Charts map[string][]RepositoryChart `json:"entries,omitempty"`
+}
+
+// FooLatest returns the latest version of charts in a Helm repository.
+func (hr *Repository) Latest() []RepositoryChart {
+	var out []RepositoryChart
+
+	for _, chart := range hr.Charts {
+		if len(chart) > 0 {
+			sort.Sort(RepositoryCharts(chart))
+			out = append(out, chart[0])
+		}
+	}
+
+	return out
+}
+
+// RepositoryChart is metadata describing a Helm Chart in a repository.
+type RepositoryChart struct {
+	Description string   `json:"description,omitempty"`
+	Name        string   `json:"name,omitempty"`
+	URLs        []string `json:"urls,omitempty"`
+	Version     string   `json:"version,omitempty"`
+}
+
+// RepositoryCharts is a slice of RepositoryChart.
+type RepositoryCharts []RepositoryChart
+
+func (rc RepositoryCharts) Len() int {
+	return len(rc)
+}
+
+func (rc RepositoryCharts) Swap(i, j int) { rc[i], rc[j] = rc[j], rc[i] }
+
+func (rc RepositoryCharts) Less(i, j int) bool {
+	v1, err := semver.Make(rc[i].Version)
+	if err != nil {
+		return false
+	}
+	v2, err := semver.Make(rc[j].Version)
+	if err != nil {
+		return false
+	}
+
+	return v1.Compare(v2) == 1
+}
+
+type Getter interface {
+	Get(string) (*http.Response, error)
+}
+
+type getter struct {
```

Comment:
I propose renaming to `httpGetter` - it will be clearer to read vs. the abstract getter interface.